src/appleseed/foundation/image/oiioimagefilewriter.cpp

```diff
@@ -0,0 +1,515 @@
+            spec.attribute("XResolution", static_cast<float>(dpm));
+            spec.attribute("YResolution", static_cast<float>(dpm));
+            spec.attribute("ResolutionUnit", "cm");
+        }
+    }
+}
+
+void OIIOImageFileWriter::set_image_attributes(const ImageAttributes& image_attributes)
+{
+    if (get_image_count() == 0)
+        throw ExceptionIOError("No images available!");
+
+    // Retrieve filename extension
+    const boost::filesystem::path filepath(m_filename);
+    const std::string extension = lower_case(filepath.extension().string());
+
+    // General image attributes
+    set_generic_image_attributes(image_attributes);
+
+    // Set image attributes depending of its extension
+    if (extension == ".exr")
+        set_exr_image_attributes(image_attributes);
+}
+
+void OIIOImageFileWriter::write_tiles(const size_t image_index)
+{
+    assert(image_index < m_images->m_canvas.size());
+
+    // Retrieves canvas
+    assert(image_index < m_images->m_canvas.size());
+    const ICanvas* canvas = m_images->m_canvas[image_index];
+    assert(canvas);
+
+    // Retrieves canvas properties
+    const CanvasProperties& props = canvas->properties();
+
+    // Retrieves image spec
+    assert(image_index < m_images->m_spec.size());
+    const OIIO::ImageSpec& spec = m_images->m_spec[image_index];
+
+    // Computes the tiles' xstride offset in bytes
+    size_t xstride = props.m_pixel_size;
+
+    // Loops over the columns of tiles
+    for (size_t tile_y = 0; tile_y < props.m_tile_count_y; tile_y++)
+    {
+        // Loops over the rows of tiles
+        for (size_t tile_x = 0; tile_x < props.m_tile_count_x; tile_x++)
+        {
+            // Computes the offset of the tile in pixels from the origin (origin: x=0;y=0).
+            const size_t tile_offset_x = tile_x * props.m_tile_width;
+            assert(tile_offset_x <= props.m_canvas_width);
+            const size_t tile_offset_y = tile_y * props.m_tile_height;
+            assert(tile_offset_y <= props.m_canvas_height);
+
+            // Computes the tile's ystride offset in bytes
+            const size_t ystride = xstride * std::min(static_cast<size_t>(spec.width + spec.x - tile_offset_x), 
+                                                      static_cast<size_t>(spec.tile_width));
+
+            // Retrieves the (tile_x, tile_y) tile.
+            const Tile& tile = canvas->tile(tile_x, tile_y);
+
+            // Writes the tile into the file.
+            if (!m_writer->write_tile(
+                    static_cast<int>(tile_offset_x),
+                    static_cast<int>(tile_offset_y),
+                    0,
+                    convert_pixel_format(props.m_pixel_format),
+                    tile.get_storage(),
+                    xstride,
+                    ystride))
+            {
+                const std::string msg = m_writer->geterror();
+                close_file();
+                throw ExceptionIOError(msg.c_str());
```

Comment:
Here throwing exception makes perfect sense.

Reply:
I don't understand. It's an exception...